### ISSUE TITLE
Question about plotting bbox on original image

### ISSUE THREAD

**Issue by nbejenne** [ultralytics/ultralytics#21279]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi !! 

### Context
I trained a model yolov8n from a custom dataset with images that I've resized manually to 640*640 using letterboxing pattern. 

This is the simple script I use to plot annotations : 

```
from ultralytics import YOLO
import cv2

model = YOLO('best.pt')
results = model('test.jpg')[0]

annotated_img = results.plot()

cv2.imwrite('result.jpg', annotated_img)
cv2.imshow('result', annotated_img)
cv2.waitKey(0)
cv2.destroyAllWindows()

```
### Problem

Now when I want to test my prediction on new images (from smartphone - with different sizes) it detects well the items but it sets a padding that make the bbox unaccurate (since I want to calculate distances between my items).

NB: The bbox are very accurate on 640*640 images. 

### Questions

Is not Yolo supposed to recalculate coordinates according to the original image (I know it does resizing by his own before process image)?

Should I retrain my model with raw images or am I missing something in the post processing workflow ? 

Thanks

**Comment by UltralyticsAssistant** [comment]
👋 Hello @nbejenne, thank you for reaching out and for providing details about your workflow! 🚀 This is an automated response to help you get started—an Ultralytics engineer will also review your question and assist you soon.

We recommend checking out the [Docs](https://docs.ultralytics.com/), especially the [Python usage](https://docs.ultralytics.com/usage/python/) and [CLI usage](https://docs.ultralytics.com/usage/cli/) sections, where many common questions and best practices are covered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) so we can take a closer look. For questions about custom training or inference, sharing example images and output visualizations can help us assist you better! You may also want to review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community in your favorite space:
- Real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth discussions: [Discourse](https://community.ultralytics.com/)
- Community threads: [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

Before proceeding, please ensure you are using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO runs seamlessly in the following verified environments (with all dependencies, including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Looking forward to your MRE or any additional details if this is a bug report. If you have further questions about post-processing or model training, please share more about your workflow and data. Thank you for helping make Ultralytics better!

**Comment by Y-T-G** [comment]
Are you referring to output from Ultralytics? The coordinates it provides should be with respect to the original input without any padding or resizing.

**Comment by nbejenne** [comment]
Exactly. 

This is the problem I face : 

![Image](https://github.com/user-attachments/assets/ea00190a-cdde-40bf-8841-8d43959b3251)

**Comment by Y-T-G** [comment]
I think you should train without manually padding the dataset images yourself. Ultralytics performs the padding as needed.

**Comment by nbejenne** [comment]
That’s what I was afraid of.. Need to annotate again.. Thank you @Y-T-G

**Comment by RizwanMunawar** [comment]
@nbejenne BTW, you can rescale the coordinates if you are using it for advanced operations.

**Comment by nbejenne** [comment]
Do you mean for re-annotating or post processing with my current setup ?

**Comment by RizwanMunawar** [comment]
I mean, if you already have the bounding box data, you can rescale it to fit on your new image. Here's the sample code.

```python
new_h, new_w = 720, 1280  # your desired size
original_h, original_w = image.shape[:2]  # original image height and width

model = YOLO('yolov8n.pt')
results = model(image)[0]  # original image

for box in results.boxes.xyxy:
    x1, y1, x2, y2 = box.cpu().numpy()

    # Rescale to target image size
    x1 = x1 * (target_w / original_w)
    x2 = x2 * (target_w / original_w)
    y1 = y1 * (target_h / original_h)
    y2 = y2 * (target_h / original_h)

    # Now x1, x2, y1, y2 are resized coordinates that you can use
```

**Comment by nbejenne** [comment]
I've already tried it but doesn't change anything since I want the new image to be the same size as the original image...

**Comment by glenn-jocher** [comment]
The issue you're experiencing is due to the double letterboxing effect - your training images were manually letterboxed to 640x640, but YOLO applies its own letterboxing during inference. This creates a mismatch between training and inference preprocessing.

The rescaling approach should work, but you need to account for the original letterboxing applied to your training data. Since your model expects the letterboxed version as "original", you'll need to reverse both the YOLO letterboxing and your manual letterboxing:

```python
from ultralytics import YOLO
import cv2

model = YOLO('best.pt')
results = model('test.jpg')[0]

# Get original image dimensions
orig_h, orig_w = results.orig_shape

# Your training images were 640x640 letterboxed versions
# Calculate the scale factor used in your manual letterboxing
scale_factor = min(640/orig_w, 640/orig_h)
new_w = int(orig_w * scale_factor)
new_h = int(orig_h * scale_factor)

# Adjust coordinates to account for your manual letterboxing
for box in results.boxes.xyxy:
    x1, y1, x2, y2 = box.cpu().numpy()
    
    # Remove the padding offsets from your manual letterboxing
    pad_x = (640 - new_w) // 2
    pad_y = (640 - new_h) // 2
    
    x1 = (x1 - pad_x) / scale_factor
    x2 = (x2 - pad_x) / scale_factor
    y1 = (y1 - pad_y) / scale_factor
    y2 = (y2 - pad_y) / scale_factor
```

However, retraining with raw images would be the cleaner long-term solution.

**Comment by nbejenne** [comment]
@glenn-jocher 
That's what I figured out !! I was able to make it accurate doing the rescaling you submitted.

I might retrain the model aswell. 

Thanks !

**Comment by glenn-jocher** [comment]
Great to hear the rescaling solution worked for you! Retraining with raw images will definitely be the cleaner approach going forward and will avoid these coordinate transformation complexities entirely.
